src/metabase/api/common.clj

```diff
@@ -329,6 +335,113 @@
+          (do
+            (try
+              (.write out (str chunk))
+              (.flush out)
+              (catch org.eclipse.jetty.io.EofException e
+                (log/info e (u/format-color 'yellow "connection closed, canceling request"))
+                (async/close! output-queue)
+                (throw e)))
+            (recur (async/<!! output-queue)))
+
+          ;; An error has occurred, let the user know
+          (instance? Exception chunk)
+          (json/generate-stream {:error (.getMessage ^Exception chunk)} out)
+
+          ;; We've recevied the response, write it to the output stream and we're done
+          (seq chunk)
+          (json/generate-stream chunk out)
+
+          ;;chunk is nil meaning the output channel has been closed
+          :else
+          out)))))
+
+(def ^:private InvokeWithKeepAliveSchema
+  {;; Channel that contains any number of newlines followed by the results of the invoked query thunk
+   :output-channel  (s/protocol async-proto/Channel)
+   ;; This channel will have an exception if that error condition is hit before the first heartbeat time, if a
+   ;; heartbeat has been sent, this channel is closed and its no longer useful
+   :error-channel   (s/protocol async-proto/Channel)
+   ;; Future that is invoking the query thunk. This is mainly useful for testing metadata to see if the future has been
+   ;; cancelled or was completed successfully
+   :response-future java.util.concurrent.Future})
+
+(s/defn ^:private invoke-thunk-with-keepalive :- InvokeWithKeepAliveSchema
+  "This function does the heavy lifting of invoking `query-thunk` on a background thread and returning it's results
+  along with a heartbeat while waiting for the results. This function returns a map that includes the relevate
+  execution information, see `InvokeWithKeepAliveSchema` for more information"
+  [query-thunk]
+  (let [response-chan (async/chan 1)
+        output-chan   (async/chan 1)
+        error-chan    (async/chan 1)
+        response-fut  (future
+                        (try
+                          (async/>!! response-chan (query-thunk))
+                          (async/close! error-chan)
+                          (catch Exception e
+                            (async/>!! error-chan e)
+                            (async/close! error-chan)
+                            (async/>!! response-chan e))))]
+    (async/go-loop []
+      (let [[response-or-timeout c] (async/alts!! [response-chan (async/timeout streaming-response-keep-alive-interval-ms)])]
+        (if response-or-timeout
+          ;; We have a response since it's non-nil, write the results and close, we're done
+          (do
+            ;; If output-chan is closed, it's already too late, nothing else we need to do
+            (async/>!! output-chan response-or-timeout)
+            (async/close! output-chan))
+          (do
+            ;; We don't have a result yet, but enough time has passed, let's assume it's not an error
```

Comment:
I don’t follow. Why has enough time passed? Ie. how doesn’t this close error-chan prematurely for response-fut?

Reply:
The error channel's only purpose is to sniff out when an exception has occurred before we send the first heartbeat string. Once we've sent that, we are officially a 200 "OK" status response. The error channel here is for something like an query syntax error, if it's an immediate error, before we've sent that first heartbeat, go ahead and bubble that exception up to the user.